homeworks/hw12/src/bank.rs

```diff
@@ -0,0 +1,999 @@
+                    self.push_transaction(operation)?;
+                    info!(
+                        "Transaction id: {} Transferred {} from {} to {}",
+                        transaction_id, amount, sender_account, receiver_account
+                    );
+                    Ok(transaction_id.to_owned())
+                }
+            } else {
+                account_not_found_error!(receiver_account.to_string())
+            }
+        } else {
+            account_not_found_error!(sender_account.to_string())
+        }
+    }
+
+    /// Returns the current balance of the account.
+    /// # Arguments
+    ///
+    /// * `account` - The code of the account for which to retrieve the balance.
+    ///
+    /// # Returns
+    /// The current balance of the account.
+    /// # Errors
+    /// AccountNotFoundError
+    /// ```
+    fn get_balance(&self, account: &str) -> Result<Money, BankError> {
+        debug!("get_balance {}", account);
+        check_account_exists!(self, account.to_string());
+        Ok(self
+            .accounts
+            .get(account)
+            .map(|balance| *balance.borrow())
+            .unwrap())
+    }
+
+    /// Returns the transaction history of the Bank.
+    ///
+    /// # Arguments
+    ///
+    /// # Returns
+    ///
+    /// A vector of [Operation] representing the transaction history of the account.
+    ///
+    /// # Errors
+    /// BankError
+    /// Returns an error if the specified account does not exist.
+    /// ```
+    fn get_history(&self) -> Result<Vec<&Operation>, BankError> {
+        let hist = self.history.iter().map(|k| k.1).collect::<Vec<_>>();
+        Ok(hist)
+    }
+
+    /// Returns the transaction history of the specified account.
+    ///
+    /// # Arguments
+    ///
+    /// * `account` - The name of the account for which to retrieve the transaction history.
+    ///
+    /// # Returns
+    ///
+    /// A vector of strings representing the transaction history of the account.
+    ///
+    /// # Errors
+    /// BankError
+    /// ```
+    fn get_account_history(&self, account: &str) -> Result<Vec<&Operation>, BankError> {
+        check_account_exists!(self, account.to_string());
+        let transaction_history = self.accounts_history.get(account);
+        let transaction_history = transaction_history.unwrap();
+        Ok(transaction_history
+            .iter()
+            .map(|t| self.history.get(t).unwrap())
+            .collect())
+    }
+
+    /// Replays the transaction history stored in a source_bank for the new Bank instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `operations_log` - history of operations to replay
+    ///
+    /// # Returns
+    /// new instance of Bank
+    /// # Errors
+    /// Returns an error if the specified account does not exist, if the file does not exist,
+    /// or if there was an error while replaying the transaction history.
+    ///
+    /// ```
+    fn replay_history(operations_log: Iter<&Operation>) -> Bank {
```

Comment:
Так будет более универсально:
```suggestion
    fn replay_history(operations_log: impl Iterator<Item = &Operation>) -> Bank {
```

А так даже еще более универсально (все, что является `Iterator`, также является и `IntoIterator`, так что можно скармливать сразу итераторы):

```suggestion 
    fn replay_history(operations_log: impl IntoIterator<Item = &Operation>) -> Bank {
```